test/mlir/onnx/onnx_lowering.mlir

```diff
@@ -2330,3 +2330,69 @@ func @test_resize1(%arg0 : tensor<3x4xf32>) -> tensor<*xf32> {
 // CHECK-DAG:       "krnl.random_normal"([[ALLOC]], [[ALL_VALUES]], [[MEAN]], [[SCALE]], [[SEED]]) : (memref<3x4x5xf64>, index, f64, f64, f64) -> ()
 // CHECK:           return [[ALLOC]] : memref<3x4x5xf64>
 }
+
+//-----
+  func @test_random_normal_like1(%arg0: tensor<3x4x5xf32>) -> tensor<*xf32> {
+    %0 = "onnx.RandomNormalLike"(%arg0) {dtype = 1 : si64, mean = 0.0 :f32, scale = 1.0 : f32, seed = 2.0 : f32} : (tensor<3x4x5xf32>) -> tensor<*xf32>
+    "std.return"(%0) : (tensor<*xf32>) -> ()
+// CHECK-LABEL:  @test_random_normal_like1
+// CHECK-DAG:       [[ALLOC:%.+]] = memref.alloc() {alignment = 16 : i64} : memref<3x4x5xf32>
+// CHECK-DAG:       [[ALL_VALUES:%.+]] = arith.constant 60 : index
+// CHECK-DAG:       [[MEAN:%.+]] = arith.constant 0.000000e+00 : f32
+// CHECK-DAG:       [[SCALE:%.+]] = arith.constant 1.000000e+00 : f32
+// CHECK-DAG:       [[SEED:%.+]] = arith.constant 2.000000e+00 : f32
+// CHECK-DAG:       "krnl.random_normal"([[ALLOC]], [[ALL_VALUES]], [[MEAN]], [[SCALE]], [[SEED]]) : (memref<3x4x5xf32>, index, f32, f32, f32) -> ()
+// CHECK:           return [[ALLOC]] : memref<3x4x5xf32>
+}
+
+//-----
+  func @test_random_normal_like2(%arg0: tensor<3x4x5xf32>) -> tensor<*xf32> {
+    %0 = "onnx.RandomNormalLike"(%arg0) {dtype = 2 : si64, mean = 0.0 :f32, scale = 1.0 : f32, seed = 2.0 : f32} : (tensor<3x4x5xf32>) -> tensor<*xf32>
+    "std.return"(%0) : (tensor<*xf32>) -> ()
+// CHECK-LABEL:  @test_random_normal_like2
+// CHECK-DAG:       [[ALLOC:%.+]] = memref.alloc() {alignment = 16 : i64} : memref<3x4x5xf64>
+// CHECK-DAG:       [[ALL_VALUES:%.+]] = arith.constant 60 : index
+// CHECK-DAG:       [[MEAN:%.+]] = arith.constant 0.000000e+00 : f64
+// CHECK-DAG:       [[SCALE:%.+]] = arith.constant 1.000000e+00 : f64
+// CHECK-DAG:       [[SEED:%.+]] = arith.constant 2.000000e+00 : f64
+// CHECK-DAG:       "krnl.random_normal"([[ALLOC]], [[ALL_VALUES]], [[MEAN]], [[SCALE]], [[SEED]]) : (memref<3x4x5xf64>, index, f64, f64, f64) -> ()
+// CHECK:           return [[ALLOC]] : memref<3x4x5xf64>
```

Comment:
This appears strange to me, the lowered code returns a memref with element type `f64`, however the function return (prior to lowering) returns a vector with element type `f32`.  I do not understand how an operator in the function body can cause the element type of the function return type to change.

Reply:
The new type comes from the `dtype` attribute.

Reply:
My understanding is that the `onnx.RandomNormalLike` operator output type is determined by the value of the `dtype` attribute (if present). In this case the test case has: 

````
%0 = "onnx.RandomNormalLike"(%arg0) {dtype = 2 : ....} : (tensor<3x4x5xf32>) -> tensor<*xf32>
````

I submit this code illegal because:
  - dtype = 2 means the output tensor element type will be `f64`
  - the operation return type is `tensor<*xf32>`
  - the declared return type has element type `f32` which not match the value of `dtype` (`f64`)
  
I think it should be:

```
 func @test_random_normal_like2(%arg0: tensor<3x4x5xf32>) -> tensor<*xf64> {
    %0 = "onnx.RandomNormalLike"(%arg0) {dtype = 2 : si64, mean = 0.0 :f32, scale = 1.0 : f32, seed = 2.0 : f32} : (tensor<3x4x5xf32>) -> tensor<*xf64>
    "std.return"(%0) : (tensor<*xf64>) -> ()
 ```

Reply:
The verification code for this operator should check that the declared output type is consistent with the value of `dtype`, and emit an error if there is a mismatch. Would be good actually to add diagnostic tests along those lines...

Reply:
I understand the point you're trying to make but in our case, unranked tensors are treated as wildcard/placeholder tensors, the type inside them is irrelevant. it's like saying tensor<*>. No checks should be performed on unranked tensors because their just placeholders with no binding type information.